allowBlank option on numericality validator

sohara opened the pull request:
I don't intend this as a real pull request but I would like to get your feedback early on how best to implement this feature. 

On the implementation: I've add a simple conditional on the `validate` method of `Ember.Validators.NumericalityValidator`. I'm not sure if this a too ugly way to do it? If I continue this way I would add this conditional to validate method of the other validators. Perhaps another way would be to implement this somehow in the parent `Ember.Validator` class. But this would meed the classes that extend `Ember.Validator` would need to all call `_super()` and then perhaps you lose the error handling to ensure they all have a `validate` function. Also, it doesn't make sense for the presence validator. Of course I'll modify the documentation for the real pull request.

On the test: It's the simplest test, just to ensure blank strings are allow through. But it passes and none of the other tests are failing as a result.

If you have a moment, please let me know your thoughts on the best approach and I'll go back and revise.

Thanks,
Sean

sohara committed:
Add allowBlank option on numericality validator with simple passing test

lcoq commented:
I think the best way to do this is to rename the `validate` methods of each validator: `_validate`. 
When someone wants to write a validator, he has to implement the `_validate` method. This method is private.

Then, each `Ember.Validator` should have a `validate` method that will call its `shouldSkipValidations(object, attr, value)` and will call `_validate` if the returned value is truthy. This method is firstly implemented in `Ember.Validator`. and just return `false` by default.

This implementation results in some changes, so if you won't have time for them, I will do it when possible.
Of course, if you have any better idea or remark, I'll be glad to read it!

sohara commented:
That sounds like a good way to do it. I'll work on it this evening or tomorrow and get you a new pull request.

sohara committed:
Initial implemention of allowBlank option by extraction validate function to _validate private function.

sohara committed:
Complete implementation of allowing optional skipping of validations.
The validate function is now part of the Ember.Validator class. It simply
calls the private _validate function if shouldSkipValidations returns
false. shouldSkipValidations returns false by default.

sohara commented:
Hi @lcoq I've add a couple of commits that change the implementation to what you suggested. I also added a note about skipping validations using the `allowBlank` option in the readme file. Let me know what you think.
Sean

sohara committed:
Update API docs to reflect change to code (Validators must implemment
_validate instead of validate).

sohara commented:
Also, let me know if you prefer that I rebase these changes into a single commit.

lcoq commented on packages/ember-validations/lib/validators/length.js:
```diff
@@ -44,5 +44,14 @@ Ember.Validators.LengthValidator = Ember.Validator.extend(/** @scope Ember.Valid
+      return false;
+    } else {
+      return true;
+    }
```
you could do something like:

``` javascript
var options = get(this, 'options');
if (options.allowBlank === true) {
  return value === ""  || value === null || value === undefined;
}
return false;
```

It might be a personal taste, but I dislike using `if (...) { return false; } else { return true; }`

lcoq commented on packages/ember-validations/lib/validators/numericality.js:
```diff
@@ -51,44 +51,47 @@ Ember.ValidationError.addMessages({
-      if (optionValue !== null && parsedValue !== optionValue) {
-        errors.add(attr, 'notEqual', {value: optionValue});
+    var options = get(this, 'options');
+    if (!((options.allowBlank === true) && ((value ==="") || (value === null) || (value === undefined)))) {
```
you can remove this check, it is already checked in the `shouldSkipValidations`

lcoq commented:
I really like what you did, except for one thing:
I told you the method `shouldSkipValidations` just return `false` by default.
After reading your code, I believe it should return `true` when the `allowBlank` is `true` by default.
So, except for the `PresenceValidator`, no one validator has to override the `shouldSkipValidations`. What do you think? Do you agree?

Anyway, thanks a lot for you time

sohara commented:
Sure, I think the things you mentioned make sense. I will try to make these changes tomorrow and let you know when it has been updated. Thanks!

sohara committed:
Implement shouldSkipValidations in Validator class.

shouldSkipValidations will return true when allowBlank option is
set to true. The method is re-implemented in PresenceValidator to
always return false.

sohara committed:
Update README to reflect changes to code regarding skipping of
validations.

sohara commented:
I've updated the pull request according to your suggestions. Let me know what you think! Thanks :)

lcoq commented:
Merged in master. Thanks @sohara !

karlguillotte commented:
Awesome, thanks guys! Nice addition to the API.
